### PR TITLE
FEATURE: Allow categories to be prioritized/prioritized in search.

### PR DESCRIPTION
**Pull request by tgxworld**

### CODE REVIEW

**Comment by tgxworld** [comment]
@SamSaffron Can you review this for me? Thanks :+1:

File: lib/search.rb [hunk 1]
```diff
@@ -837,8 +837,26 @@ def posts_query(limit, opts = nil)
...
+          #{ts_query(weight_filter: weights)},
+          0|32
+        ) *
+        (
+          CASE categories.search_priority
+          WHEN #{Searchable::PRIORITIES[:very_low]}
+          THEN #{SiteSetting.category_search_priority_very_low_weight}
+          WHEN #{Searchable::PRIORITIES[:low]}
+          THEN #{SiteSetting.category_search_priority_low_weight}
+          WHEN #{Searchable::PRIORITIES[:normal]} THEN 1
```
Comment (ZogStriP) [inline lib/search.rb]: Do you need this line when it's the default and covered by the `ELSE`?

File: lib/validators/category_search_priority_weights_validator.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      val < 1 && val > SiteSetting.category_search_priority_very_low_weight
+    when "category_search_priority_high_weight"
+      val > 1 && val < SiteSetting.category_search_priority_very_high_weight
+    when "category_search_priority_very_high_weight"
+      val > SiteSetting.category_search_priority_high_weight
+    end
+  end
+
+  def error_message
+    key =
```
Comment (ZogStriP) [inline lib/validators/category_search_priority_weights_validator.rb]: You could also do 

``` ruby
key = @name[/category_search_priority_(\w+)_weight/, 1]
```

File: lib/search.rb [hunk 3]
```diff
@@ -837,8 +837,26 @@ def posts_query(limit, opts = nil)
       end
     else
       data_ranking = <<~SQL
-      TS_RANK_CD(
-        post_search_data.search_data, #{ts_query(weight_filter: weights)}
+      (
+        TS_RANK_CD(
+          post_search_data.search_data,
+          #{ts_query(weight_filter: weights)},
+          0|32
```
Comment (ZogStriP) [inline lib/search.rb]: These magic numbers (flags) need at the very least a comment explaining what they do

Comment (tgxworld) [reply lib/search.rb]: Hmm these numbers are in the documentation for the function being used here and I would prefer not to have to duplicate the Postgres docs here.

Comment (ZogStriP) [reply lib/search.rb]: I don't mean to copy-paste the postgres doc, but why do you use this combination of flags?

File: config/site_settings.yml [hunk 4]
```diff
@@ -1614,6 +1614,18 @@ search:
       ro: true
       sk: true
       tr_TR: true
+  category_search_priority_very_low_weight:
```
Comment (SamSaffron) [inline config/site_settings.yml]: I am uneasy with all these toggles in the UX it feels like something that will just cause confusion to admins, can we start off with just having these as hidden site settings?

**Comment by SamSaffron** [comment]
feel free to merge once my comment is addressed :+1:

**Comment by tgxworld** [comment]
Thanks for reviewing @ZogStriP and @SamSaffron

**Pull request merged**